workflow/controller/taskresult.go

```diff
@@ -83,4 +90,5 @@ func (woc *wfOperationCtx) taskResultReconciliation() {
 			woc.updated = true
 		}
 	}
+	return artifactsWriting
 }
```

Comment:
Is it possible for the TaskResult to have been created and not yet have that label on it and get to here?

Reply:
The task result is upserted [here](https://github.com/argoproj/argo-workflows/blob/f9e026a660b0f4f1f4aca2c7022c00370c555fde/workflow/executor/executor.go#L297) with the label before artifacts are written, so no, I don't think so.

Reply:
Sorry, trying to understand. So, if this method gets called before the label is added and before any artifacts are written, then this method will return false, right? Then "false" will be passed into `garbageCollectArtifacts()`, right? (probably missing something but not sure what yet) :)

Reply:
Yes, if this method gets called before the label is added and before artifacts are written, this method will return false. If any artifacts are in the process of being written, the `common.LabelKeyWritingArtifact` label will be `true`, which will prevent garbage collection for at least this loop through the controller.

Reply:
`common.LabelKeyWritingArtifact` is set to false once `reportResult` of executor.go happens.

Reply:
@juliev0 Sorry, I misunderstood your original question. Yes, it is possible for the TaskResult to have been created without the label (in `monitorProgress` of executor.go`).

Reply:
So, continuing with the scenario where this method gets called before the label is ever added and then returns false, we will enter into `processArtifactGCCompletion()`, right? Isn't it possible to prematurely remove the finalizer?

Reply:
`processArtifactGCCompletion` still checks `removeFinalizer = anyPodSuccess && woc.allArtifactsDeleted()`. The problem before was that there was an unreported artifact (due to the race condition), which allowed the finalizer to be removed. The new label makes it so that we don't enter `processArtifactGCCompletion` if we know an artifact hasn't been reported yet.

Reply:
Hmm... it might be possible that it would prematurely remove the finalizer if the reconciliation and GC happens before an artifact is even attempted to be written... E.g. if one wait container takes a long time for some reason?

If that's the case, perhaps it would be safer to add something like the following at the beginning of the wait container before `wfExecutor.Wait(ctx)`?:

```
func (we *WorkflowExecutor) ExpectArtifact(ctx context.Context) error {
	if len(we.Template.Outputs.Artifacts) == 0 {
		return nil
	}

	// Create an empty task result with a label to indicate that artifacts are to be written. Prevents garbage collection race condition.
	err = we.reportResult(ctx, wfv1.NodeResult{}, map[string]string{common.LabelKeyWritingArtifact: "true"})
	if err != nil {
		return err
	}
}
```

Logically, this would imply "don't process artifact GC completion until all the wait containers with artifacts to write are done.

Reply:
`we.monitorProgress` (the only other instance where `reportResult` is called), is called in `wfExecutor.Wait(ctx)`, so the solution above would ensure the label is always set when there are artifacts to be written.

Reply:
Or even better, we can put the above in the the `init` container? I assume the init container always runs before the `main` container?

Reply:
What if instead of the current Label, we have something simple which indicates whether the TaskResult has been fully written, and when the TraskResult gets created it starts off with that label set to false?

Reply:
I can implement what you're suggesting as it might be easier to understand. I'll make it so that when the wait container starts, it creats an empty task result with an unfinished label, and then just before it exits, I'll switch the label to true. In reconciliation, all results must be marked finished, and then `processArtifactGCCompletion()` can proceed.

How does this sound?

Reply:
Yes, that would be great. Thank you. I'm thinking there could be some `defer` at the top of `waitContainer()` which will make it so this label gets set true even if we prematurely exit that method for some reason. And that would call some method of `wfExecutor`. Is this what you're thinking?

Reply:
Due to the nature of reconciliation + the fact that wait containers finish at different times, I found it necessary to set up some state in `wfOperationCtx` which tracks completions on a per task-result basis, maintaining completion status between reconciliation runs.

Previously, It was possible that reconciliation could execute and there only be one task-result in the list of objects, even if there were multiple artifacts in the workflow. This would cause false positives for "all complete" logic since it looks as though all have completed but only one had actually completed.

I've modified the test case to increase the frequency of the race condition (though it's not 100%, it's better than ~1/30). I'm running a battery of tests now.

Reply:
There’s one task result per workflow pod that executes I think, right? Or is it one task result per Workflow?

Reply:
Per pod.

Reply:
"Previously, It was possible that reconciliation could execute and there only be one task-result in the list of objects, even if there were multiple artifacts in the workflow." - that could happen for a Workflow running just one Pod which has multiple output artifacts, right?

Reply:
Perhaps once your commit is in, I can see what you mean better...

Reply:
Sorry, my comment was very unclear. "previously" was the wrong word. It can happen with one pod and multiple artifacts. E,g. the user-defined artifact as well as the container log artifact. I'm not really sure how it's happening with the log artifact + user-defined artifact for a given container since they are reported at the same time in the `wait` container (`outputs.Artifacts = append(outputs.Artifacts, logArtifacts...)` The `SharedInformer` cache's must not represent state changes of the entire target object all at once? By extension, it happens with multiple user-defined artifacts across multiple pods. 

I had to modify the state I mentioned above, as `wfOperationCtx` doesn't persist across each operator loop. I ended up putting it in with `ArtifactGCStatus`.

I will push shortly.

Reply:
Ok, still getting intermittent errors, but I think I found the issue:

```
time="2023-10-25T15:37:21.967Z" level=info msg="TaskSet Reconciliation" namespace=argo workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:21.967Z" level=info msg=reconcileAgentPod namespace=argo workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:21.967Z" level=info msg="Updated phase Running -> Failed" namespace=argo workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:21.967Z" level=info msg="Updated message  -> Stopped with strategy 'Stop'" namespace=argo workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:21.967Z" level=info msg="Marking workflow completed" namespace=argo workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:21.967Z" level=debug msg="Event(v1.ObjectReference{Kind:\"Workflow\", Namespace:\"argo\", Name:\"artgc-dag-wf-stopped-wfwgk\", UID:\"c9e57468-6cb8-4f95-8b97-29908ac7015d\", APIVersion:\"argoproj.io/v1alpha1\", ResourceVersion:\"619562\", FieldPath:\"\"}): type: 'Warning' reason: 'WorkflowFailed' Stopped with strategy 'Stop'"
time="2023-10-25T15:37:21.968Z" level=debug msg="Log changes patch: {\"metadata\":{\"labels\":{\"workflows.argoproj.io/completed\":\"true\",\"workflows.argoproj.io/phase\":\"Failed\"}},\"status\":{\"conditions\":[{\"status\":\"True\",\"type\":\"PodRunning\"},{\"status\":\"True\",\"type\":\"Completed\"}],\"finishedAt\":\"2023-10-25T15:37:21Z\",\"message\":\"Stopped with strategy 'Stop'\",\"nodes\":{\"artgc-dag-wf-stopped-wfwgk\":{\"finishedAt\":\"2023-10-25T15:37:21Z\",\"outboundNodes\":[\"artgc-dag-wf-stopped-wfwgk-3359801380\",\"artgc-dag-wf-stopped-wfwgk-3058715464\"],\"phase\":\"Failed\"},\"artgc-dag-wf-stopped-wfwgk-3058715464\":{\"finishedAt\":\"2023-10-25T15:37:21Z\",\"message\":\"workflow shutdown with strategy:  Stop\",\"phase\":\"Failed\"},\"artgc-dag-wf-stopped-wfwgk-3359801380\":{\"finishedAt\":\"2023-10-25T15:37:21Z\",\"message\":\"workflow shutdown with strategy:  Stop\",\"phase\":\"Failed\"}},\"phase\":\"Failed\"}}"
time="2023-10-25T15:37:21.968Z" level=info msg="Workflow to be dehydrated" Workflow Size=5874
time="2023-10-25T15:37:21.972Z" level=info msg="cleaning up pod" action=terminateContainers key=argo/artgc-dag-wf-stopped-wfwgk-artgc-dag-artifact-creator-2-3359801380/terminateContainers
time="2023-10-25T15:37:21.972Z" level=info msg="cleaning up pod" action=deletePod key=argo/artgc-dag-wf-stopped-wfwgk-1340600742-agent/deletePod
time="2023-10-25T15:37:21.972Z" level=info msg="https://0.0.0.0:42429/api/v1/namespaces/argo/pods/artgc-dag-wf-stopped-wfwgk-artgc-dag-artifact-creator-2-3359801380/exec?command=%2Fvar%2Frun%2Fargo%2Fargoexec&command=kill&command=15&command=1&container=main&stderr=true&stdout=true&tty=false"
time="2023-10-25T15:37:21.973Z" level=debug msg="exec container command" command="[/var/run/argo/argoexec kill 15 1]" container=main error="<nil>" namespace=argo pod=artgc-dag-wf-stopped-wfwgk-artgc-dag-artifact-creator-2-3359801380
time="2023-10-25T15:37:21.972Z" level=info msg="cleaning up pod" action=terminateContainers key=argo/artgc-dag-wf-stopped-wfwgk-artgc-dag-workflow-stopper-3058715464/terminateContainers
time="2023-10-25T15:37:21.973Z" level=info msg="https://0.0.0.0:42429/api/v1/namespaces/argo/pods/artgc-dag-wf-stopped-wfwgk-artgc-dag-workflow-stopper-3058715464/exec?command=%2Fvar%2Frun%2Fargo%2Fargoexec&command=kill&command=15&command=1&container=main&stderr=true&stdout=true&tty=false"
time="2023-10-25T15:37:21.973Z" level=debug msg="exec container command" command="[/var/run/argo/argoexec kill 15 1]" container=main error="<nil>" namespace=argo pod=artgc-dag-wf-stopped-wfwgk-artgc-dag-workflow-stopper-3058715464
time="2023-10-25T15:37:21.973Z" level=debug msg="Create events 201"
time="2023-10-25T15:37:21.978Z" level=debug msg="Update workflows 200"
time="2023-10-25T15:37:21.979Z" level=info msg="Workflow update successful" namespace=argo phase=Failed resourceVersion=619565 workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:21.980Z" level=debug msg="Event(v1.ObjectReference{Kind:\"Workflow\", Namespace:\"argo\", Name:\"artgc-dag-wf-stopped-wfwgk\", UID:\"c9e57468-6cb8-4f95-8b97-29908ac7015d\", APIVersion:\"argoproj.io/v1alpha1\", ResourceVersion:\"619565\", FieldPath:\"\"}): type: 'Normal' reason: 'WorkflowNodeRunning' Running node artgc-dag-wf-stopped-wfwgk.stop-workflow: workflow shutdown with strategy:  Stop"
time="2023-10-25T15:37:21.980Z" level=debug msg="Event(v1.ObjectReference{Kind:\"Workflow\", Namespace:\"argo\", Name:\"artgc-dag-wf-stopped-wfwgk\", UID:\"c9e57468-6cb8-4f95-8b97-29908ac7015d\", APIVersion:\"argoproj.io/v1alpha1\", ResourceVersion:\"619565\", FieldPath:\"\"}): type: 'Warning' reason: 'WorkflowNodeFailed' Failed node artgc-dag-wf-stopped-wfwgk.stop-workflow: workflow shutdown with strategy:  Stop"
time="2023-10-25T15:37:21.980Z" level=debug msg="Event(v1.ObjectReference{Kind:\"Workflow\", Namespace:\"argo\", Name:\"artgc-dag-wf-stopped-wfwgk\", UID:\"c9e57468-6cb8-4f95-8b97-29908ac7015d\", APIVersion:\"argoproj.io/v1alpha1\", ResourceVersion:\"619565\", FieldPath:\"\"}): type: 'Warning' reason: 'WorkflowNodeFailed' Failed node artgc-dag-wf-stopped-wfwgk.create-artifact-2: workflow shutdown with strategy:  Stop"
time="2023-10-25T15:37:21.980Z" level=debug msg="Event(v1.ObjectReference{Kind:\"Workflow\", Namespace:\"argo\", Name:\"artgc-dag-wf-stopped-wfwgk\", UID:\"c9e57468-6cb8-4f95-8b97-29908ac7015d\", APIVersion:\"argoproj.io/v1alpha1\", ResourceVersion:\"619565\", FieldPath:\"\"}): type: 'Warning' reason: 'WorkflowNodeFailed' Failed node artgc-dag-wf-stopped-wfwgk"
time="2023-10-25T15:37:21.981Z" level=debug msg="Delete pods 404"
time="2023-10-25T15:37:21.994Z" level=debug msg="Create events 201"
time="2023-10-25T15:37:22.002Z" level=debug msg="Create events 201"
time="2023-10-25T15:37:22.011Z" level=debug msg="Create events 201"
time="2023-10-25T15:37:22.017Z" level=debug msg="Create events 201"
time="2023-10-25T15:37:22.019Z" level=debug msg="DeleteCollection workflowtaskresults 200"
time="2023-10-25T15:37:22.025Z" level=info msg="cleaning up pod" action=labelPodCompleted key=argo/artgc-dag-wf-stopped-wfwgk-artgc-dag-workflow-stopper-3058715464/labelPodCompleted
time="2023-10-25T15:37:22.025Z" level=info msg="cleaning up pod" action=labelPodCompleted key=argo/artgc-dag-wf-stopped-wfwgk-artgc-dag-artifact-creator-2-3359801380/labelPodCompleted
time="2023-10-25T15:37:22.036Z" level=debug msg="Patch pods 200"
time="2023-10-25T15:37:22.038Z" level=debug msg="Patch pods 200"
time="2023-10-25T15:37:22.058Z" level=debug msg="Create pods/exec 500"
time="2023-10-25T15:37:22.058Z" level=info msg="signaled container" container=main error="unable to upgrade connection: container not found (\"main\")" namespace=argo pod=artgc-dag-wf-stopped-wfwgk-artgc-dag-workflow-stopper-3058715464 stderr="<nil>" stdout="<nil>"
time="2023-10-25T15:37:22.058Z" level=info msg="https://0.0.0.0:42429/api/v1/namespaces/argo/pods/artgc-dag-wf-stopped-wfwgk-artgc-dag-workflow-stopper-3058715464/exec?command=%2Fvar%2Frun%2Fargo%2Fargoexec&command=kill&command=15&command=1&container=wait&stderr=true&stdout=true&tty=false"
time="2023-10-25T15:37:22.058Z" level=debug msg="exec container command" command="[/var/run/argo/argoexec kill 15 1]" container=wait error="<nil>" namespace=argo pod=artgc-dag-wf-stopped-wfwgk-artgc-dag-workflow-stopper-3058715464
time="2023-10-25T15:37:22.059Z" level=debug msg="Create pods/exec 101"
time="2023-10-25T15:37:22.078Z" level=debug msg="Create pods/exec 101"
time="2023-10-25T15:37:22.127Z" level=info msg="signaled container" container=wait error="command terminated with exit code 137" namespace=argo pod=artgc-dag-wf-stopped-wfwgk-artgc-dag-workflow-stopper-3058715464 stderr="<nil>" stdout="<nil>"
time="2023-10-25T15:37:22.148Z" level=info msg="signaled container" container=main error="<nil>" namespace=argo pod=artgc-dag-wf-stopped-wfwgk-artgc-dag-artifact-creator-2-3359801380 stderr= stdout="killing 1 with terminated\n"
time="2023-10-25T15:37:22.148Z" level=info msg="https://0.0.0.0:42429/api/v1/namespaces/argo/pods/artgc-dag-wf-stopped-wfwgk-artgc-dag-artifact-creator-2-3359801380/exec?command=%2Fvar%2Frun%2Fargo%2Fargoexec&command=kill&command=15&command=1&container=wait&stderr=true&stdout=true&tty=false"
time="2023-10-25T15:37:22.148Z" level=debug msg="exec container command" command="[/var/run/argo/argoexec kill 15 1]" container=wait error="<nil>" namespace=argo pod=artgc-dag-wf-stopped-wfwgk-artgc-dag-artifact-creator-2-3359801380
time="2023-10-25T15:37:22.164Z" level=debug msg="Create pods/exec 101"
time="2023-10-25T15:37:22.311Z" level=info msg="signaled container" container=wait error="<nil>" namespace=argo pod=artgc-dag-wf-stopped-wfwgk-artgc-dag-artifact-creator-2-3359801380 stderr= stdout="killing 1 with terminated\n"
time="2023-10-25T15:37:22.969Z" level=debug msg="Syncing all CronWorkflows"
time="2023-10-25T15:37:23.040Z" level=info msg="Processing workflow" namespace=argo workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:23.041Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=1 workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:23.041Z" level=debug msg="task result:\n&WorkflowTaskResult{ObjectMeta:{artgc-dag-wf-stopped-wfwgk-3058715464  argo  2caee164-7b62-4046-86ce-b19838b09efe 619575 1 2023-10-25 15:37:22 +0000 UTC <nil> <nil> map[workflows.argoproj.io/report-outputs-completed:true workflows.argoproj.io/workflow:artgc-dag-wf-stopped-wfwgk] map[] [{v1 Pod artgc-dag-wf-stopped-wfwgk-artgc-dag-workflow-stopper-3058715464 6bcf6446-89a2-4a7b-a1b9-3c546e06cbec <nil> <nil>}] []  [{argoexec Update argoproj.io/v1alpha1 2023-10-25 15:37:22 +0000 UTC FieldsV1 {\"f:metadata\":{\"f:labels\":{\".\":{},\"f:workflows.argoproj.io/report-outputs-completed\":{},\"f:workflows.argoproj.io/workflow\":{}},\"f:ownerReferences\":{\".\":{},\"k:{\\\"uid\\\":\\\"6bcf6446-89a2-4a7b-a1b9-3c546e06cbec\\\"}\":{}}},\"f:outputs\":{\".\":{},\"f:artifacts\":{}}} }]},NodeResult:NodeResult{Phase:,Message:,Outputs:&Outputs{Parameters:[]Parameter{},Artifacts:[]Artifact{Artifact{Name:main-logs,Path:,Mode:nil,From:,ArtifactLocation:ArtifactLocation{ArchiveLogs:nil,S3:&S3Artifact{S3Bucket:S3Bucket{Endpoint:,Bucket:,Region:,Insecure:nil,AccessKeySecret:nil,SecretKeySecret:nil,RoleARN:,UseSDKCreds:false,CreateBucketIfNotPresent:nil,EncryptionOptions:nil,CASecret:nil,},Key:artgc-dag-wf-stopped-wfwgk/artgc-dag-wf-stopped-wfwgk-artgc-dag-workflow-stopper-3058715464/main.log,},Git:nil,HTTP:nil,Artifactory:nil,HDFS:nil,Raw:nil,OSS:nil,GCS:nil,Azure:nil,},GlobalName:,Archive:nil,Optional:false,SubPath:,RecurseMode:false,FromExpression:,ArtifactGC:nil,Deleted:false,},},Result:nil,ExitCode:nil,},Progress:,},}" namespace=argo workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:23.041Z" level=debug msg="task result name:\nartgc-dag-wf-stopped-wfwgk-3058715464" namespace=argo workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:23.041Z" level=debug msg="task result counter:\nmap[artgc-dag-wf-stopped-wfwgk-1114725310:0 artgc-dag-wf-stopped-wfwgk-3058715464:0 artgc-dag-wf-stopped-wfwgk-3359801380:1 artgc-dag-wf-stopped-wfwgk-3410134237:0]" namespace=argo workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:23.041Z" level=debug msg="task result counter len: 4" namespace=argo workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:23.041Z" level=info msg="task-result changed" namespace=argo nodeID=artgc-dag-wf-stopped-wfwgk-3058715464 workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:23.041Z" level=debug msg="Skipping artifact GC" namespace=argo workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:23.042Z" level=debug msg="Log changes patch: {\"status\":{\"nodes\":{\"artgc-dag-wf-stopped-wfwgk-3058715464\":{\"outputs\":{\"artifacts\":[{\"name\":\"main-logs\",\"s3\":{\"key\":\"artgc-dag-wf-stopped-wfwgk/artgc-dag-wf-stopped-wfwgk-artgc-dag-workflow-stopper-3058715464/main.log\"}}]}}}}}"
time="2023-10-25T15:37:23.042Z" level=info msg="Workflow to be dehydrated" Workflow Size=6064
time="2023-10-25T15:37:23.237Z" level=debug msg="Update workflows 200"
time="2023-10-25T15:37:23.237Z" level=info msg="Workflow update successful" namespace=argo phase=Failed resourceVersion=619577 workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:23.258Z" level=debug msg="DeleteCollection workflowtaskresults 200"
time="2023-10-25T15:37:23.419Z" level=debug msg="Get leases 200"
time="2023-10-25T15:37:23.426Z" level=debug msg="Update leases 200"
time="2023-10-25T15:37:24.242Z" level=info msg="Processing workflow" namespace=argo workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:24.243Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=0 workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:24.243Z" level=debug msg="Skipping artifact GC" namespace=argo workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:25.128Z" level=info msg="cleaning up pod" action=killContainers key=argo/artgc-dag-wf-stopped-wfwgk-artgc-dag-workflow-stopper-3058715464/killContainers
time="2023-10-25T15:37:25.312Z" level=info msg="cleaning up pod" action=killContainers key=argo/artgc-dag-wf-stopped-wfwgk-artgc-dag-artifact-creator-2-3359801380/killContainers
```

Important lines are:
```
time="2023-10-25T15:37:22.127Z" level=info msg="signaled container" container=wait error="command terminated with exit code 137" namespace=argo pod=artgc-dag-wf-stopped-wfwgk-artgc-dag-workflow-stopper-3058715464 stderr="<nil>" stdout="<nil>"
```
and
```
time="2023-10-25T15:37:23.041Z" level=debug msg="task result counter:\nmap[artgc-dag-wf-stopped-wfwgk-1114725310:0 artgc-dag-wf-stopped-wfwgk-3058715464:0 artgc-dag-wf-stopped-wfwgk-3359801380:1 artgc-dag-wf-stopped-wfwgk-3410134237:0]" namespace=argo workflow=artgc-dag-wf-stopped-wfwgk
time="2023-10-25T15:37:23.041Z" level=debug msg="task result counter len: 4" namespace=argo workflow=artgc-dag-wf-stopped-wfwgk
```

The wait container is being killed when the test fails.

I'm thinking it is to do with the following not being at the top level of the wait container:
```
ctx := context.Background() // don't allow cancellation to impact capture of results, parameters, or artifacts
```
`NewWaitCommand` looks like this:
```
func NewWaitCommand() *cobra.Command {
	command := cobra.Command{
		Use:   "wait",
		Short: "wait for main container to finish and save artifacts",
		Run: func(cmd *cobra.Command, args []string) {
			ctx := cmd.Context()
			err := waitContainer(ctx)
			if err != nil {
				log.Fatalf("%+v", err)
			}
		},
	}
	return &command
}
```
I think `cmd.Context()` should be `ctx.Background()`, no?

Reply:
Hmm, still not it.

```
time="2023-10-25T17:49:35.706Z" level=debug msg="exec container command" command="[/var/run/argo/argoexec kill 15 1]" container=main error="<nil>" namespace=argo pod=artgc-dag-wf-stopped-pxhfh-artgc-dag-artifact-creator-2-428993330
time="2023-10-25T17:49:35.706Z" level=debug msg="exec container command" command="[/var/run/argo/argoexec kill 15 1]" container=main error="<nil>" namespace=argo pod=artgc-dag-wf-stopped-pxhfh-artgc-dag-workflow-stopper-2323127074
time="2023-10-25T17:49:35.709Z" level=debug msg="Create events 201"
time="2023-10-25T17:49:35.709Z" level=info msg="cleaning up pod" action=deletePod key=argo/artgc-dag-wf-stopped-pxhfh-1340600742-agent/deletePod
time="2023-10-25T17:49:35.714Z" level=debug msg="Update workflows 200"
time="2023-10-25T17:49:35.715Z" level=info msg="Workflow update successful" namespace=argo phase=Failed resourceVersion=634290 workflow=artgc-dag-wf-stopped-pxhfh
time="2023-10-25T17:49:35.715Z" level=debug msg="Event(v1.ObjectReference{Kind:\"Workflow\", Namespace:\"argo\", Name:\"artgc-dag-wf-stopped-pxhfh\", UID:\"66ecac64-54a8-4f40-bf89-ec795ab52ad3\", APIVersion:\"argoproj.io/v1alpha1\", ResourceVersion:\"634290\", FieldPath:\"\"}): type: 'Warning' reason: 'WorkflowNodeFailed' Failed node artgc-dag-wf-stopped-pxhfh"
time="2023-10-25T17:49:35.715Z" level=debug msg="Event(v1.ObjectReference{Kind:\"Workflow\", Namespace:\"argo\", Name:\"artgc-dag-wf-stopped-pxhfh\", UID:\"66ecac64-54a8-4f40-bf89-ec795ab52ad3\", APIVersion:\"argoproj.io/v1alpha1\", ResourceVersion:\"634290\", FieldPath:\"\"}): type: 'Normal' reason: 'WorkflowNodeRunning' Running node artgc-dag-wf-stopped-pxhfh.stop-workflow: workflow shutdown with strategy:  Stop"
time="2023-10-25T17:49:35.715Z" level=debug msg="Event(v1.ObjectReference{Kind:\"Workflow\", Namespace:\"argo\", Name:\"artgc-dag-wf-stopped-pxhfh\", UID:\"66ecac64-54a8-4f40-bf89-ec795ab52ad3\", APIVersion:\"argoproj.io/v1alpha1\", ResourceVersion:\"634290\", FieldPath:\"\"}): type: 'Warning' reason: 'WorkflowNodeFailed' Failed node artgc-dag-wf-stopped-pxhfh.stop-workflow: workflow shutdown with strategy:  Stop"
time="2023-10-25T17:49:35.715Z" level=debug msg="Event(v1.ObjectReference{Kind:\"Workflow\", Namespace:\"argo\", Name:\"artgc-dag-wf-stopped-pxhfh\", UID:\"66ecac64-54a8-4f40-bf89-ec795ab52ad3\", APIVersion:\"argoproj.io/v1alpha1\", ResourceVersion:\"634290\", FieldPath:\"\"}): type: 'Warning' reason: 'WorkflowNodeFailed' Failed node artgc-dag-wf-stopped-pxhfh.create-artifact-2: workflow shutdown with strategy:  Stop"
time="2023-10-25T17:49:35.720Z" level=info msg="Queueing Failed workflow argo/artgc-dag-wf-stopped-dhvxd for delete due to max rention(2 workflows)"
time="2023-10-25T17:49:35.720Z" level=info msg="Deleting garbage collected workflow 'argo/artgc-dag-wf-stopped-dhvxd'"
time="2023-10-25T17:49:35.728Z" level=debug msg="Delete pods 404"
time="2023-10-25T17:49:35.733Z" level=debug msg="Create events 201"
time="2023-10-25T17:49:35.744Z" level=debug msg="Delete workflows 404"
time="2023-10-25T17:49:35.744Z" level=info msg="Workflow already deleted 'argo/artgc-dag-wf-stopped-dhvxd'"
time="2023-10-25T17:49:35.747Z" level=debug msg="Create events 201"
time="2023-10-25T17:49:35.754Z" level=debug msg="Create events 201"
time="2023-10-25T17:49:35.758Z" level=debug msg="Create pods/exec 101"
time="2023-10-25T17:49:35.759Z" level=debug msg="Create pods/exec 101"
time="2023-10-25T17:49:35.767Z" level=debug msg="Create events 201"
time="2023-10-25T17:49:35.772Z" level=debug msg="DeleteCollection workflowtaskresults 200"
time="2023-10-25T17:49:35.778Z" level=info msg="cleaning up pod" action=labelPodCompleted key=argo/artgc-dag-wf-stopped-pxhfh-artgc-dag-workflow-stopper-2323127074/labelPodCompleted
time="2023-10-25T17:49:35.778Z" level=info msg="cleaning up pod" action=labelPodCompleted key=argo/artgc-dag-wf-stopped-pxhfh-artgc-dag-artifact-creator-2-428993330/labelPodCompleted
time="2023-10-25T17:49:35.791Z" level=debug msg="Patch pods 200"
time="2023-10-25T17:49:35.791Z" level=debug msg="Patch pods 200"
time="2023-10-25T17:49:35.819Z" level=info msg="signaled container" container=main error="Internal error occurred: error executing command in container: failed to exec in container: failed to create exec \"c7d9abff3b18c09bb17e19d4c45a6c48256fc5b536c383043e5b683ff7f4c1f1\": task 06a801f5c6bdf2576ffa32494b941be68fa3a0e30e0ef5f15b893e1632d7f46c not found: not found" namespace=argo pod=artgc-dag-wf-stopped-pxhfh-artgc-dag-workflow-stopper-2323127074 stderr="<nil>" stdout="<nil>"
time="2023-10-25T17:49:35.819Z" level=info msg="https://0.0.0.0:42429/api/v1/namespaces/argo/pods/artgc-dag-wf-stopped-pxhfh-artgc-dag-workflow-stopper-2323127074/exec?command=%2Fvar%2Frun%2Fargo%2Fargoexec&command=kill&command=15&command=1&container=wait&stderr=true&stdout=true&tty=false"
time="2023-10-25T17:49:35.819Z" level=debug msg="exec container command" command="[/var/run/argo/argoexec kill 15 1]" container=wait error="<nil>" namespace=argo pod=artgc-dag-wf-stopped-pxhfh-artgc-dag-workflow-stopper-2323127074
time="2023-10-25T17:49:35.866Z" level=info msg="signaled container" container=main error="<nil>" namespace=argo pod=artgc-dag-wf-stopped-pxhfh-artgc-dag-artifact-creator-2-428993330 stderr= stdout="killing 1 with terminated\n"
time="2023-10-25T17:49:35.866Z" level=info msg="https://0.0.0.0:42429/api/v1/namespaces/argo/pods/artgc-dag-wf-stopped-pxhfh-artgc-dag-artifact-creator-2-428993330/exec?command=%2Fvar%2Frun%2Fargo%2Fargoexec&command=kill&command=15&command=1&container=wait&stderr=true&stdout=true&tty=false"
time="2023-10-25T17:49:35.867Z" level=debug msg="exec container command" command="[/var/run/argo/argoexec kill 15 1]" container=wait error="<nil>" namespace=argo pod=artgc-dag-wf-stopped-pxhfh-artgc-dag-artifact-creator-2-428993330
time="2023-10-25T17:49:35.867Z" level=debug msg="Create pods/exec 101"
time="2023-10-25T17:49:35.881Z" level=info msg="signaled container" container=wait error="Internal error occurred: error executing command in container: failed to exec in container: failed to create exec \"81ed5a77b9b27168c3fa79c2d23eee476dfd30bd48107dbb4c8135a10109cf10\": cannot exec in a stopped state: unknown" namespace=argo pod=artgc-dag-wf-stopped-pxhfh-artgc-dag-workflow-stopper-2323127074 stderr="<nil>" stdout="<nil>"
time="2023-10-25T17:49:35.887Z" level=debug msg="Create pods/exec 101"
time="2023-10-25T17:49:36.017Z" level=info msg="signaled container" container=wait error="<nil>" namespace=argo pod=artgc-dag-wf-stopped-pxhfh-artgc-dag-artifact-creator-2-428993330 stderr= stdout="killing 1 with terminated\n"
```

Reply:
It turns out that with the new usage of `finalizeOutput` in the wait container `defer`, the `context.Background()` is now necessary at the top level, or the wait container will be killed before the defer runs.